### PR TITLE
fix(router): validate lazy loaded configs

### LINKED ISSUE #25431: Router: strict route validation, but not for lazy-loaded modules
## I'm submitting a...

<pre><code>[X] Bug report  
[X] Documentation issue or request
</code></pre>

## Current behavior

When a `Route` only contains a CanActivate guard:
```typescript
const routes = [
  {
    path: '',
    canActivate: [MyRedirectGuard] // always does a dynamic redirect
  }
];
```
You will get a validation error: 
```
Error: Invalid configuration of route ''. One of the following must be provided: component, redirectTo, children or loadChildren
```

However, when this route is in a lazy-loaded module it works perfectly fine.

## Expected behavior
I expect routes in lazy-loaded modules to have the same validation. However, I wish this validation rule to be less strict, because we use CanActivate guards for dynamic redirects.

If the rule is valid, there should be another way to have dynamic redirects on component-less routes, e.g. a new type of guard:

```typescript
const routes = [
  {
    path: '',
    redirect: MyRedirectGuard
  }
];

interface Redirect {
  redirect(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): string | Promise<string> | Observable<string>;
}

class MyRedirectGuard implements Redirect {
  redirect(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): string {
    return 'some-dynamic-url';
  }
}
```

## Minimal reproduction of the problem with instructions
https://stackblitz.com/edit/component-less-routes

Uncomment the first route in app.module.ts to trigger the error. However, in feature.module.ts it works fine.

## What is the motivation / use case for changing the behavior?

We use `CanActivate` guards to do dynamic redirects whenever we need to. We only want them to trigger for specific urls, so we have component-less routes:

```typescript
const routes = [
  {
    path: 'feature',
    component: FeatureComponent,
    children: [
      {
        path: '',
        canActivate: [MyRedirectGuard] // checks if we need to go to foo or bar
      },
      {
        path: 'foo',
        component: FooComponent
      },
      {
        path: 'bar',
        component: BarComponent
      }
    ]
  }
];
```
With this construction, when a user navigates to `/feature`, it automatically goes to `/feature/foo` or `/feature/bar`.

We want to be able to do the same in modules that are not lazy-loaded, without the need to provide some kind om "dummy" component.

## Environment

<pre><code>Angular version: 6.1.0</code></pre>

### PR DESCRIPTION
**Pull request by atscott**
Lazy loaded configs are not validated at runtime like the initial set of
routes are. This change also validates lazy loaded configs right after
they're loaded.

BREAKING CHANGE: Lazy loaded configs are now also validated once loaded like the
initial set of routes are. Lazy loaded modules which have invalid Route
configs will now error. Note that this is only done in dev mode so
there is no production impact of this change.

Fixes #25431

### CODE REVIEW

**Comment by atscott** [comment]
[presubmit](http://tap/439444098)

**Comment by atscott** [comment]
[TGP](https://fusion2.corp.google.com/presubmit/tap/439444098/OCL:439444098:BASE:439596931:1649175111533:f448eef5;groups=PossiblyNewlyFailing/targets) revealed several failures that will need to be cleaned up

**Comment by atscott** [comment]
[Green TGP](https://fusion2.corp.google.com/presubmit/tap/439444098/OCL:439444098:BASE:441783585:1649954351412:ec68ec13;groups=PossiblyNewlyFailing/targets)

File: packages/router/src/router_config_loader.ts [hunk 1]
```diff
@@ -47,15 +47,15 @@ export class RouterConfigLoader {
             this.onLoadEndListener(route);
           }
           const module = factory.create(parentInjector);
+          const routes =
+              flatten(module.injector.get(ROUTES, [], InjectFlags.Self | InjectFlags.Optional))
+                  .map(standardizeConfig);
+          validateConfig(routes);
           // When loading a module that doesn't provide `RouterModule.forChild()` preloader
```
Comment (AndrewKushnir) [inline packages/router/src/router_config_loader.ts:54]: It looks like this comment belongs to the code that was moved above?

File: packages/router/src/router_config_loader.ts [hunk 2]
```diff
@@ -47,15 +47,15 @@ export class RouterConfigLoader {
             this.onLoadEndListener(route);
           }
           const module = factory.create(parentInjector);
+          const routes =
+              flatten(module.injector.get(ROUTES, [], InjectFlags.Self | InjectFlags.Optional))
+                  .map(standardizeConfig);
+          validateConfig(routes);
```
Comment (AndrewKushnir) [inline packages/router/src/router_config_loader.ts:53]: Can we do it in dev-mode only, i.e. add the `ngDevMode` check?

**Review by AndrewKushnir — APPROVED** [review]
👍

File: packages/router/src/router_config_loader.ts (hunk 2 above)
Comment (atscott) [reply packages/router/src/router_config_loader.ts]: 👍  I also added this to the `resetConfig` which ended up ensuring the function gets removed from the prod bundle

**Comment by atscott** [comment]
merge assistance: the legacy unit test failure is unrelated. Also, you may want to merge and sync this on its own (in case there were regressions in g3 between the TGP run and now)

**Comment by dylhunn** [comment]
This PR was merged into the repository by commit 96fd29c6d2d2abc5afee4d21a3e964a79aa39844.

**Pull request closed without merging**